include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,692 @@
+/*@{*/
+
+/**
+ * A class for storing at each cell @p CellIteratorType a vector of data @p DataType .
+ * The underlying structure and the initialize() method of this class is designed
+ * in such a way that one could use different child class derived from the
+ * base DataType class to store data on a given cell. This implies the usage of pointers, in our
+ * case -- std::shared_ptr().
+ *
+ * @note The limitation is that the data stored on each cell should be of the same type.
+ * For this reason, this class may not be sufficiently
+ * flexible when, for example, adopting a level-set approach to describe material behavior.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CellIteratorType, typename DataType>
+class CellDataStorage
+{
+public:
+  /**
+   * A typedef for a vector of shared pointers with arbitrary data type @p T .
+   */
+  template <typename T> using CellVector = std::vector<std::shared_ptr<T> >;
+
+  /**
+   * Constructor.
+   */
+  CellDataStorage();
+
+  /**
+   * Destructor.
+   */
+  virtual ~CellDataStorage();
+
+  /**
+   * Initialize data on the @p cell to store @p n objects of type @p T .
+   * The typename @p T is possibly another class which is derived from the
+   * base @p DataType class. In order to initialize the vector of objects
+   * we have to assume that the class @p T has a standard constructor.
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell,
+                  const unsigned int n);
+
+  /**
+   * Get a vector of the data located at @p cell .
+   * A possible additional typename @p T is the class to which the
+   * base class DataType should be casted.
+   * Since @p DataType is stored as shared pointers, there is minimal overhead in
+   * returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on the
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<T> get_data(const CellIteratorType &cell);
+
+  /**
+   * Get const pointers to data located at @p cell .
+   * A possible additional typename @p T is the class to which the
+   * base class DataType should be casted.
+   * Since DataType is stored as shared pointers, there is minimal overhead in
+   * returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on the
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<const T> get_data(const CellIteratorType &cell) const;
+
+private:
+  /**
+   * A map to store a vector of data at the cell.
+   */
+  std::map<CellIteratorType, CellVector<DataType> > map;
+
+};
+
+
+/**
+ * An abstract class which specifies requirements for a data class on
+ * a quadrature point to be transferable via restriction/prolongation matrices
+ * or other mechanisms.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+class TransferableQuadratureData
```

Comment:
@tjhei : would you prefer `TransferableQuadraturePointData` as a name to stress this?